wreck max walltime support

grondo opened the pull request:
This PR adds advisory timeout support to wreck jobs in the form of a new `timeout.lua` script under `lua.d`. If `lwj.<id>.walltime` exists the plugin starts a timer after the job reaches the `running` state. When the timer fires the job is killed with `SIGALRM`.

Additionally a `-T, --walltime` option is added to `wreck.lua` which takes walltime option in `NUMBER[SUFFIX]` form where `NUMBER` can be arbitrary floating point number and optional suffix is "s" for seconds (default), "m" for minutes, "h" for hours, or "d" for days (like GNU `sleep(1)`).

A default of `0` is used to indicate no timeout.

@surajpkn: Can you rebase #409 on top of this PR, and make any comments?

Example:
```
grondo@flux-core:~/workspace (wreck-timeout) $ flux wreckrun -N4 --walltime=.5m sleep 60
[1443326159.790466] job.info[0]: got request job.create
[1443326159.791379] job.info[0]: Setting job 2 to reserved
[1443326160.113298] lwj.2.info[1]: lwj 2: node1: nprocs=1, nnodes=4, cmdline=[ "sleep", "60" ]
[1443326160.144830] lwj.2.info[0]: lwj 2: node0: nprocs=1, nnodes=4, cmdline=[ "sleep", "60" ]
[1443326160.158572] lwj.2.info[3]: lwj 2: node3: nprocs=1, nnodes=4, cmdline=[ "sleep", "60" ]
[1443326160.322712] lwj.2.info[2]: lwj 2: node2: nprocs=1, nnodes=4, cmdline=[ "sleep", "60" ]
[1443326160.477934] lwj.2.info[0]: detected job state = running, starting timer for 30s
[1443326190.501886] lwj.2.info[0]: Timeout expired! Killing job
[1443326190.502335] lwj.2.info[0]: Killing jobid 2 with signal 14
[1443326190.502503] lwj.2.info[3]: Killing jobid 2 with signal 14
[1443326190.502735] lwj.2.info[1]: Killing jobid 2 with signal 14
[1443326190.503852] lwj.2.info[2]: Killing jobid 2 with signal 14
[1443326190.508051] lwj.2.info[2]: job complete. exiting...
[1443326190.510384] lwj.2.info[0]: job complete. exiting...
[1443326190.508217] lwj.2.info[3]: job complete. exiting...
[1443326190.514647] lwj.2.info[1]: job complete. exiting...
wreckrun: tasks [0-3]: exited with signal 14
[1443326190.556522] job.info[0]: got request job.disconnect
```

lipari commented on src/modules/wreck/lua.d/timeout.lua:
```diff
@@ -0,0 +1,40 @@
+        oneshot = true,
+        handler = function (f, to)
+          wreck:log_msg ("Timeout expired! Killing job")
+          local rc,err = f:sendevent ({signal = 14}, "wreck.%d.kill", id)
```
Wouldn't it be better to use signal name (SIGALRM) instead of number?  Also, I expected SIGTERM instead of SIGALRM.

grondo replied:
We could, but that would require loading posix module which otherwise isn't needed.

I'm not tied to SIGALRM, but it mimics `run_timeout` and `alarm(3)` behavior, and also return code 142 can be checked for jobs terminated in this way, whereas SIGTERM may be used for jobs that cancelled?

Easy to change though.

lipari replied:
Agreed w/r//t using a number.  I understand the case for signal 14, but it's breaking with tradition - whether that's good or bad...

grondo replied:
I think you're right on the signal name thing. It is pretty simple to load posix module anyway. I'll push a change for your review (maybe this evening)

For SIGALRM vs SIGTERM, maybe at this point (since this is still in the prototype phase) we could just make it configurable?

lipari replied:
That works!

grondo commented:
Ok, I've just pushed some updates as discussed with @lipari. I kept the new commits separate for now, so they're easier to review, but will rebase and squash things together before merging.

@lipari: Similar to environment handling, the wreck timeout.lua allows the default `walltime-signal` (SIGALRM) to be overridden either per session, by setting `lwj.walltime-signal`, or per job via `lwj.<id>.walltime-signal`. The signal can be set either as a number (e.g. `15`) or name (e.g. `SIGTERM`).

I've added tests for this functionality, and also fixed a small bug in the original timeout test.

lipari commented:
@grondo, very nice!

grondo commented:
Thanks for the review @lipari!

I squashed up some commits here and rebased, so I think this one is ready to go. (perhaps pending @surajpkn's ACK that this approach will work for him.)

garlick commented:
Ready to push the button after ack from @surajpkn

surajpkn commented:
Sorry for the delay. This is great!
